### ISSUE TITLE
Enable/Test embedding ops

### ISSUE THREAD

**Issue by shruthip-venkatesh** [torch-spyre/torch-spyre#3336]
### What feature would you like to be added?

Currently, `aten.embedding.default` is a CPU fallback. Enable it to natively run on Spyre.

### Why is this feature needed?

This feature is needed for Embedding Lookup

**Comment by shruthip-venkatesh** [comment]
Merged the PR with the fix.

**Comment by moriohara** [comment]
@shruthip-venkatesh toch.nn.functional.embedding is failing for the following use cases:

| Test Case | Status | Tensor Type | Tensor Name | Shape | Stride | DType |
|-----------|--------|-------------|-------------|-------|--------|-------|
| test_model_ops_db_torch_nn_functional_embedding_1__Ministral-3-14B-Instruct-2512_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 14] | [14, 1] | i64 |
|  |  | INPUT | arg1_1 | [131072, 5120] | [5120, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 14, 5120] | [71680, 5120, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_1__Mistral-Small-3_2-24B-Instruct-2506_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 855] | [855, 1] | i64 |
|  |  | INPUT | arg1_1 | [131072, 5120] | [5120, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 855, 5120] | [4377600, 5120, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_1__Qwen2_5-7B-Instruct_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 39] | [39, 1] | i64 |
|  |  | INPUT | arg1_1 | [152064, 3584] | [3584, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 39, 3584] | [139776, 3584, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_1__gpt-oss-20b_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 11] | [11, 1] | i64 |
|  |  | INPUT | arg1_1 | [201088, 2880] | [2880, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 11, 2880] | [31680, 2880, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__Meta-Llama-3_1-8B-Instruct_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [128256, 4096] | [4096, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 4096] | [4096, 4096, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__Ministral-3-14B-Instruct-2512_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [131072, 5120] | [5120, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 5120] | [5120, 5120, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__Mistral-Small-3_2-24B-Instruct-2506_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [131072, 5120] | [5120, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 5120] | [5120, 5120, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__Qwen2_5-7B-Instruct_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [152064, 3584] | [3584, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 3584] | [3584, 3584, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__gpt-oss-20b_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [201088, 2880] | [2880, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 2880] | [2880, 2880, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__granite-3_3-8b-instruct-fms_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [49159, 4096] | [4096, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 4096] | [4096, 4096, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__granite-3_3-8b-instruct_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [49159, 4096] | [4096, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 4096] | [4096, 4096, 1] | bf16 |
| test_model_ops_db_torch_nn_functional_embedding_2__granite-4_1-8b_yaml__spyre_float16 | FAILED | INPUT | arg0_1 | [1, 1] | [1, 1] | i64 |
|  |  | INPUT | arg1_1 | [100352, 4096] | [4096, 1] | bf16 |
|  |  | OUTPUT | embedding | [1, 1, 4096] | [4096, 4096, 1] | bf16 |

**Comment by arielge** [comment]
Adding to @moriohara's comment above, I am seeing several embedding-related accuracy issues for end-to-end model runs in HF adapters. The code compiles and runs, but silently returns incorrect values once the embedding source layout crosses the per-core tensor-span limit.

These are the results of a shape sweep:
| Weight shape | Representative checkpoints | Compiler-reported best per-core span | Result |
|---|---|---:|---:|
| `(149760, 3584)` | Synthetic below-boundary control | below 256 MiB | PASS, max abs `6.10e-05` |
| `(149888, 3584)` | Synthetic first failing boundary point | 256.156 MiB | FAIL, max abs `0.10666` |
| `(151936, 896)` | `Qwen/Qwen2-0.5B` | below 256 MiB | PASS, max abs `6.10e-05` |
| `(151936, 2048)` | `Qwen/Qwen2.5-3B-Instruct` | below 256 MiB | PASS, max abs `6.10e-05` |
| `(152064, 3584)` | `Qwen/Qwen2.5-7B-Instruct` | 259.875 MiB | FAIL, max abs `0.10666` |
| `(152064, 5120)` | `Qwen/Qwen2.5-32B-Instruct`, `deepseek-ai/DeepSeek-R1-Distill-Qwen-14B` | 371.250 MiB | FAIL, max abs `0.10666` |
| `(131072, 4096)` | `mistralai/Ministral-3-8B-Base-2512` | 256.000 MiB | PASS, max abs `6.10e-05` |
| `(131200, 4096)` | Synthetic first failing point above 256 MiB | 256.250 MiB | FAIL, max abs `0.11157` |
| `(151936, 4096)` | `Qwen/Qwen3-8B` | 296.750 MiB | FAIL, max abs `0.11157` |
| `(100352, 4096)` | `ibm-granite/granite-4.1-8b` | below 256 MiB | PASS, max abs `6.10e-05` |
| `(49159, 4096)` | `ibm-granite/granite-3.3-8b-instruct` | below 256 MiB | PASS, max abs `6.10e-05` |

The listed geometries have the same PASS/FAIL verdict in fp16 and their configured bf16 dtype.

Reproducer code:

```python
import torch

VOCAB_SIZE = 152064
HIDDEN_SIZE = 3584
DTYPE = torch.float16
THRESHOLD = 1e-3


torch.manual_seed(20260810)

input_ids = torch.tensor([[0]], dtype=torch.int64)
weight = torch.empty(VOCAB_SIZE, HIDDEN_SIZE, dtype=DTYPE)
torch.nn.init.normal_(weight, mean=0.0, std=0.02)

cpu_output = torch.nn.functional.embedding(input_ids, weight)
cpu_direct = weight[input_ids]
cpu_control_max_abs = (cpu_output.float() - cpu_direct.float()).abs().max().item()
assert cpu_control_max_abs == 0.0

weight_spyre = weight.to("spyre")
input_ids_spyre = input_ids.to("spyre")
spyre_output = torch.nn.functional.embedding(input_ids_spyre, weight_spyre).to("cpu")

abs_error = (spyre_output.float() - cpu_output.float()).abs()
max_abs = abs_error.max().item()
over_threshold_count = torch.count_nonzero(abs_error > THRESHOLD).item()
passed = max_abs <= THRESHOLD

print(f"torch={torch.__version__}")
print(
    f"weight_shape={tuple(weight.shape)} ids_shape={tuple(input_ids.shape)} "
    f"dtype={DTYPE}"
)
print(f"cpu_control_max_abs={cpu_control_max_abs:.8g}")
print(f"max|spyre-cpu|={max_abs:.8g}")
print(f"over_threshold_count={over_threshold_count}/{abs_error.numel()}")
print(f"threshold={THRESHOLD:.8g}")
print("PASS" if passed else "FAIL")

if not passed:
    raise SystemExit(1)
```

Observed output:

```text
torch=2.13.0+cu130
weight_shape=(152064, 3584) ids_shape=(1, 1) dtype=torch.float16
cpu_control_max_abs=0
max|spyre-cpu|=0.10665894
over_threshold_count=495/3584
threshold=0.001
FAIL
```

Environment:

- torch-spyre commit: `3fd2f420d1d02915a6d2e981237c2aa3a101d7f9`
- PyTorch: `2.13.0+cu130`
- device mode: PF

Related broader coverage tracker: #2483.

**Comment by tanushree-ibm** [comment]
Here is my observation for the above test failure for aten.embedding op
**Summery:**
The test is failing : 495 out of 3584 elements have error > 1e-3, with a max error of ~0.107. That is far too large to be a floating-point rounding issue. Something is gathering the wrong data.

**Analysis:**
The key difference between the passing tests and the failing script is the VOCAB_SIZE=152064 — far larger than any test case (the largest is V=4096).

**Root Cause: Multi-Stick Embedding Dimension Not Tested**
The failing script uses weight: [152064, 3584] fp16. At 64 elements/stick:

**Embedding dim: 3584 / 64 = 56 sticks** → device_size for this axis is [56, 64] (outer count + stick)
**Vocab axis:** 152064 rows → device layout is [1, 152064, 56, 64] (4-D)
The largest test passing today is test_single_idx_embedding_large_vocab with V=4096, E=512 → 512/64 = 8 sticks, still 4-D device layout — but it only checks op-spec structure, not numerical correctness (expect=GATHER_OP_SPEC but expect_close is not True).

When **aten.embedding** compiles for a large vocab (**V=152064, E=3584**), two ops are scheduled:



Op                   Role
buf1                Restickify copy of the weight table — plain Pointwise, direct reads, device_size=[152064, 56, 64]

buf0                Gather output — indirect reads buf1 at i2 + 3584*tmp0 where tmp0 is runtime-loaded

**span_reduction_pass** runs on both ops. For **buf0** the existing **indirect_access_pinned_vars** constraint fires (it has indirect reads) and pins all dims to **split=1**. But **buf1** is a plain direct-copy op — indirect_info_from_op(buf1) returns nothing — so **span_reduction_pass** freely computes splits=**{d0:4, d1:7}** to try to fit the **152064×3584** tensor **within the 256 MB per-core span limit.** Those splits shard the weight table into **28 per-core** fragments. The gather consumer then reads buf1 with a runtime index that addresses the full vocab — so for any token whose row lives in a shard other than core 0's, the wrong data is returned. This produced max|spyre-cpu| ≈ 0.107, with 495/3584 elements wrong.


This is a known single core issue for large tensors. It can be resolved using multicore and will be fixed as a part of below PR
[https://github.com/torch-spyre/torch-spyre/pull/2699](https://github.com/torch-spyre/torch-spyre/pull/2699)
